### DISCUSSION TITLE
How to cross-compile the ksh?

### DISCUSSION THREAD

**Discussion by Hyperkopite** [ksh93/ksh#578] [Q&A]
Is there any ez way like using `./configure --host=xxx` while we compiling other softwares?

**Comment by McDutchie** [comment]
No, because the build depends on various feature tests that have to be run on the architecture and OS that the binary is going to be run on. For example, a test is done not just if `posix_spawn`(2) exists, but also to [determine if it's functional enough](https://github.com/ksh93/ksh/blob/4ec937455617514d5cb12aef4be69e7438a83232/src/lib/libast/features/lib#L233-L345) to be worth using instead of `fork`(2). There are many other such tests, some written as C programs and some as shell scripts, all embedded in the iffe test scripts.

AT&T originally tried to make cross-compilation possible by providing a `crossexec` script, invoked by iffe, that remotely runs architecture-dependent feature tests on a host of the target architecture via rsh or ssh. But, in my view, that's a bit pointless. if you have to run feature tests on a remote machine, you might as well run the build on that remote machine to begin with, and avoid the hassle of setting up automated command execution via ssh.

The crossexec script is still part of the 93u+m distro. I've actually never tried this out, so I don't know how well it works (if at all). It's just never reached the top of my priority list.

**Reply by Hyperkopite** [reply]
wow thanks a lot for this detailed  explanation. So If I'm understanding this correctly, the cross compilation is still possible by using this `crossexec`? It would be great if you can provide me with some guidance on how to use this tool!

**Comment by phidebian** [comment]
Thanks @McDutchie  for the thorough explanations. xcompile may be usefull sometimes, it happen the poor starving programmer don't have access to the 'real' remote machine and but instead have access to a qemu system instance that could be order of magnitude slower than the bare metal. With your explanations, I understand that iffe that I see when doing build actually mean (if feature OK). May be a way to remote run the iffe suites and getting back the cached iffe, could be enough to after this to xcompile, for same OS for which a multilib is available (i.e linux).

Another use case could be a target system (same os/arch) that is so tiny (embeded) that it don't host a compiler suite, but would enjoy a shell that would be more a busybox then, and ksh93 with its builtin dynload could be a good candidate.

May be a side project for volunteer, because I understand it is not a high prio...
